### PR TITLE
arithmetic on bn128 elliptic curve

### PR DESCRIPTION
**Pull request by ayrat555**

### CODE REVIEW

File: lib/bn/bn128_arithmetic.ex [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+      true ->
+        {:ok, add_points(point1, point2)}
+    end
+  end
+
+  @spec mult(Point.t(), integer(), integer()) :: {:ok, Point.t()} | {:error, String.t()}
+  def mult(point, scalar, b \\ @default_b) do
+    if !on_curve?(point, b) do
+      {:error, "point1 is not on the curve"}
```
Comment (masonforest) [inline lib/bn/bn128_arithmetic.ex]: `point1` -> `point` ?

File: lib/bn/bn128_arithmetic.ex [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+      %Point{
+        x: x,
+        y: y,
+        modulus: point1.modulus
+      }
+    end
+  end
+
+  @spec infinity?(Point.t()) :: boolean()
+  def infinity?(point) do
```
Comment (masonforest) [inline lib/bn/bn128_arithmetic.ex]: curious: why is `(0,0)` called `infinity` not `zero`?

File: lib/bn/integer_mod_p.ex [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  end
+
+  def sub(_, _) do
+    raise ArgumentError, message: "#{__MODULE__}.sub/2 can only substract #{__MODULE__} structs"
+  end
+
+  @spec mult(t(), t() | integer()) :: t()
+  def mult(number1 = %__MODULE__{}, number2 = %__MODULE__{}) do
+    if number1.modulus != number2.modulus,
+      do: raise(ArgumentError, message: "Numbers calculated with different modulus")
```
Comment (masonforest) [inline lib/bn/integer_mod_p.ex]: could you put this in a guard clause?

File: lib/bn/bn128_arithmetic.ex [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    cond do
+      scalar == 0 ->
+        {:ok, result} = Point.new(0, 0, modulus: point.modulus)
+
+        result
+
+      scalar == 1 ->
+        point
+
+      rem(scalar, 2) == 0 ->
```
Comment (masonforest) [inline lib/bn/bn128_arithmetic.ex]: Would `Integer.is_even?` work here?

File: lib/bn/integer_mod_p.ex [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    mult(number1, inverse)
+  end
+
+  def div(_, _) do
+    raise ArgumentError,
+      message: "#{__MODULE__}.sub/2 can only divide #{__MODULE__} structs"
+  end
+
+  @spec pow(t(), t()) :: t()
+  def pow(number1 = %__MODULE__{}, number2) do
```
Comment (masonforest) [inline lib/bn/integer_mod_p.ex]: Would it make sense to rename `number1` to `base` and `number2` to `exponent`?

File: lib/bn/integer_mod_p.ex [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    cond do
+      number2 == 0 ->
```
Comment (masonforest) [inline lib/bn/integer_mod_p.ex]: can this be a `case` on `number2`?

File: lib/bn/bn128_arithmetic.ex (hunk 2 above)
Comment (ayrat555) [reply lib/bn/bn128_arithmetic.ex]: @masonforest  it's described in eip196
`Curve points are encoded as two field elements (x, y), where the point at infinity is encoded as (0, 0).`

**Comment by ayrat555** [comment]
@masonforest  I fixed all your comments. Thanks for the review, Mason!

**Pull request merged**